rsocket/framing/FrameTransport.cpp

```diff
@@ -31,7 +31,7 @@ void FrameTransport::connect() {
   }
 
   connectionOutput_ = connection_->getOutput();
-  connectionOutput_->onSubscribe(yarpl::get_ref(this));
+  connectionOutput_->onSubscribe(this->get_ref(this));
```

Comment:
`get_ref_from_this()`

why do you need to give `this` to `get_ref` function if you already call it on `this`

Reply:
C++ doesn't have covariant return types; eg the return type of `get_ref` can't be deduced to anything but `enable_get_ref`, so we have to provide the subclass type by way of passing in `this`.

We _could_ have `enable_get_ref` be parameterized over the class inheriting from it, but this doesn't work well when a further derived class wants to call `get_this`; you end up with calls like:

```cpp
class MyFooBase : public enable_get_ref<MyFooBase> {};
class MyBarBase : public enable_get_ref<MyBarBase> {};

class Derived : MyFooBase, MyBarBase {
  void doSomething() {
    auto self = this->template yarpl::enable_this<MyFooBase>get_ref(); // :(
  }
};
```

which just isn't going to fly. The same problem pops up when a parent class `Bar` inherits from `enable_get_ref`, and a child class `Foo` wants to get a `Reference<Foo>`. 

So we do something which is at least consistent, which is (at runtime, on dev builds) verify that in `enable_get_ref::get_ref`, `this` can be downcasted to the concrete type of `ptr`, and check that the caller of `get_ref` is indeed getting a caller of this instance. We basically enforce the semantics of `shared_from_this` this way. 

I didn't rename this to `get_ref_from_this` because it's pretty explicit what you're getting a reference to, and that'd be three instance of the word `this` per line.